x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -150,103 +143,15 @@ public void execute(
-
-        LogicalPlanActionListener(
-            EsqlQueryRequest request,
-            EsqlExecutionInfo executionInfo,
-            BiConsumer<PhysicalPlan, ActionListener<Result>> runPhase,
-            ActionListener<Result> listener
-        ) {
-            this.request = request;
-            this.executionInfo = executionInfo;
-            this.runPhase = runPhase;
-            this.listener = listener;
-        }
-
-        @Override
-        public void onResponse(LogicalPlan analyzedPlan) {
-            executeOptimizedPlan(request, executionInfo, runPhase, optimizedPlan(analyzedPlan), listener);
-        }
-
-        /**
-         * Whether to return an empty result (HTTP status 200) for a CCS rather than a top level 4xx/5xx error.
-         *
-         * For cases where field-caps had no indices to search and the remotes were unavailable, we
-         * return an empty successful response (200) if all remotes are marked with skip_unavailable=true.
-         *
-         * Note: a follow-on PR will expand this logic to handle cases where no indices could be found to match
-         * on any of the requested clusters.
-         */
-        private boolean returnSuccessWithEmptyResult(Exception e) {
-            if (executionInfo.isCrossClusterSearch() == false) {
-                return false;
-            }
-
-            if (e instanceof NoClustersToSearchException || ExceptionsHelper.isRemoteUnavailableException(e)) {
-                for (String clusterAlias : executionInfo.clusterAliases()) {
-                    if (executionInfo.isSkipUnavailable(clusterAlias) == false
-                        && clusterAlias.equals(RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY) == false) {
-                        return false;
-                    }
-                }
-                return true;
-            }
-            return false;
-        }
-
-        @Override
-        public void onFailure(Exception e) {
-            if (returnSuccessWithEmptyResult(e)) {
-                executionInfo.markEndQuery();
-                Exception exceptionForResponse;
-                if (e instanceof ConnectTransportException) {
-                    // when field-caps has no field info (since no clusters could be connected to or had matching indices)
-                    // it just throws the first exception in its list, so this odd special handling is here is to avoid
-                    // having one specific remote alias name in all failure lists in the metadata response
-                    exceptionForResponse = new RemoteTransportException(
-                        "connect_transport_exception - unable to connect to remote cluster",
-                        null
-                    );
+            ActionListener.wrap(plan -> executeOptimizedPlan(request, executionInfo, runPhase, optimizedPlan(plan), listener), e -> {
+                if (EsqlSessionCCSUtils.returnSuccessWithEmptyResult(executionInfo, e)) {
```

Comment:
I wonder if we could go further and put this logic into `EsqlSessionCCSUtils` too. So we'd have just a method name here. After all, creating an empty result is also a generic function that is not very specific to EsqlSession?